# Lab | Hypothesis Testing

**Objective**

Welcome to the Hypothesis Testing Lab, where we embark on an enlightening journey through the realm of statistical decision-making! In this laboratory, we delve into various scenarios, applying the powerful tools of hypothesis testing to scrutinize and interpret data.

From testing the mean of a single sample (One Sample T-Test), to investigating differences between independent groups (Two Sample T-Test), and exploring relationships within dependent samples (Paired Sample T-Test), our exploration knows no bounds. Furthermore, we'll venture into the realm of Analysis of Variance (ANOVA), unraveling the complexities of comparing means across multiple groups.

So, grab your statistical tools, prepare your hypotheses, and let's embark on this fascinating journey of exploration and discovery in the world of hypothesis testing!

**Challenge 1**

In this challenge, we will be working with pokemon data. The data can be found here:

- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/pokemon.csv

In [9]:
#libraries
import pandas as pd
import scipy.stats as st
from scipy.stats import shapiro
import numpy as np
from scipy.stats import levene
from scipy.stats import ttest_ind



In [2]:
df = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/pokemon.csv")
df

,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
0,Bulbasaur,Grass,Poison,45,49,49,65,65,45,1,False
1,Ivysaur,Grass,Poison,60,62,63,80,80,60,1,False
2,Venusaur,Grass,Poison,80,82,83,100,100,80,1,False
3,Mega Venusaur,Grass,Poison,80,100,123,122,120,80,1,False
4,Charmander,Fire,NaN,39,52,43,60,50,65,1,False
...,...,...,...,...,...,...,...,...,...,...,...
795,Diancie,Rock,Fairy,50,100,150,100,150,50,6,True
796,Mega Diancie,Rock,Fairy,50,160,110,160,110,110,6,True
797,Hoopa Confined,Psychic,Ghost,80,110,60,150,130,70,6,True
798,Hoopa Unbound,Psychic,Dark,80,160,60,170,130,80,6,True


- We posit that Pokemons of type Dragon have, on average, more HP stats than Grass. Choose the propper test and, with 5% significance, comment your findings.

In [3]:
# The appropriate test is an independent two-sample t-test (Two Sample T-Test)
# Extracting HP values
dragon_hp = df[df["Type 1"] == "Dragon"]["HP"]
grass_hp = df[df["Type 1"] == "Grass"]["HP"]

In [4]:
print("Dragon Pokémon:", len(dragon_hp))
print("Grass Pokémon:", len(grass_hp))

print("\nMean HP")
print("Dragon:", dragon_hp.mean())
print("Grass:", grass_hp.mean())

Dragon Pokémon: 32
Grass Pokémon: 70

Mean HP
Dragon: 83.3125
Grass: 67.27142857142857


In [6]:
# Checking normality with Shapiro-Wilk test
dragon_shapiro = shapiro(dragon_hp)
grass_shapiro = shapiro(grass_hp)

print("Dragon:", dragon_shapiro)
print("Grass:", grass_shapiro)

Dragon: ShapiroResult(statistic=np.float64(0.9635281236696637), pvalue=np.float64(0.3419890019248819))
Grass: ShapiroResult(statistic=np.float64(0.9715915656401263), pvalue=np.float64(0.11214356726520774))


Both p-values are greater than 0.05.
There is no evidence that either group's HP values are not normally distributed.

In [8]:
stat, p = levene(dragon_hp, grass_hp)

print("Levene's test statistic:", stat)
print("p-value:", p)

Levene's test statistic: 1.867317074967163
p-value: 0.1748473993433114


0.175 > 0.05
There is no evidence that the variances are different.

In [10]:
# Dragon HP > Grass HP, we perform a one-tailed independent t-test.
t_stat, p_two_tailed = ttest_ind(dragon_hp, grass_hp, equal_var=True)

# Convert to one-tailed p-value
p_one_tailed = p_two_tailed / 2

print("t-statistic:", t_stat)
print("Two-tailed p-value:", p_two_tailed)
print("One-tailed p-value:", p_one_tailed)

t-statistic: 3.590444254130357
Two-tailed p-value: 0.0005135938300306967
One-tailed p-value: 0.0002567969150153484


Findings:
An independent two-sample t-test was used because we compared the mean HP of two independent groups (Dragon and Grass Pokémon). The assumptions were checked and satisfied: both groups were approximately normally distributed (Shapiro-Wilk, p > 0.05) and had equal variances (Levene's test, p = 0.175). The t-test produced a statistic of t = 3.590 with a one-tailed p-value = 0.000257. Since the p-value is less than the significance level of 0.05, we reject the null hypothesis and conclude that Dragon-type Pokémon have a significantly higher average HP than Grass-type Pokémon.

- We posit that Legendary Pokemons have different stats (HP, Attack, Defense, Sp.Atk, Sp.Def, Speed) when comparing with Non-Legendary. Choose the propper test and, with 5% significance, comment your findings.


In [12]:
#Two-Sample T-Test
legendary = df[df["Legendary"] == True]
non_legendary = df[df["Legendary"] == False]

In [13]:
print("Legendary:", len(legendary))
print("Non-Legendary:", len(non_legendary))

Legendary: 65
Non-Legendary: 735


In [14]:
stats = ["HP", "Attack", "Defense", "Sp. Atk", "Sp. Def", "Speed"]

for stat in stats:
    t_stat, p_value = ttest_ind(
        legendary[stat],
        non_legendary[stat],
        equal_var=False
    )
    
    print(stat)
    print("t-statistic:", t_stat)
    print("p-value:", p_value)
    print("-------------------")

HP
t-statistic: 8.981370483625046
p-value: 1.0026911708035289e-13
-------------------
Attack
t-statistic: 10.438133539322203
p-value: 2.520372449236657e-16
-------------------
Defense
t-statistic: 7.637078164784618
p-value: 4.8269984949193406e-11
-------------------
Sp. Atk
t-statistic: 13.417449984138461
p-value: 1.5514614112239793e-21
-------------------
Sp. Def
t-statistic: 10.015696613114878
p-value: 2.2949327864052933e-15
-------------------
Speed
t-statistic: 11.47504444631443
p-value: 1.049016311882455e-18
-------------------


In [15]:
for stat in stats:
    print(stat)
    print("Legendary mean:", legendary[stat].mean())
    print("Non-Legendary mean:", non_legendary[stat].mean())
    print()

HP
Legendary mean: 92.73846153846154
Non-Legendary mean: 67.18231292517007

Attack
Legendary mean: 116.67692307692307
Non-Legendary mean: 75.66938775510204

Defense
Legendary mean: 99.66153846153846
Non-Legendary mean: 71.55918367346939

Sp. Atk
Legendary mean: 122.18461538461538
Non-Legendary mean: 68.45442176870748

Sp. Def
Legendary mean: 105.93846153846154
Non-Legendary mean: 68.89251700680272

Speed
Legendary mean: 100.18461538461538
Non-Legendary mean: 65.45578231292517



Findings:
Independent two-sample t-tests were conducted to compare the stats of Legendary and Non-Legendary Pokémon. The analysis showed statistically significant differences for all evaluated attributes (HP, Attack, Defense, Sp. Atk, Sp. Def, and Speed), with all p-values below the 5% significance level. Therefore, we reject the null hypothesis and conclude that Legendary Pokémon have significantly different stats compared with Non-Legendary Pokémon. By comparing the group means, we observe that Legendary Pokémon have higher average values across all six attributes.

**Challenge 2**

In this challenge, we will be working with california-housing data. The data can be found here:
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/california_housing.csv

In [16]:
df = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/california_housing.csv")
df.head()

,longitude,latitude,housing_median_age,total_rooms,total_bedrooms,population,households,median_income,median_house_value
0,-114.31,34.19,15.0,5612.0,1283.0,1015.0,472.0,1.4936,66900.0
1,-114.47,34.40,19.0,7650.0,1901.0,1129.0,463.0,1.8200,80100.0
2,-114.56,33.69,17.0,720.0,174.0,333.0,117.0,1.6509,85700.0
3,-114.57,33.64,14.0,1501.0,337.0,515.0,226.0,3.1917,73400.0
4,-114.57,33.57,20.0,1454.0,326.0,624.0,262.0,1.9250,65500.0


**We posit that houses close to either a school or a hospital are more expensive.**

- School coordinates (-118, 34)
- Hospital coordinates (-122, 37)

We consider a house (neighborhood) to be close to a school or hospital if the distance is lower than 0.50.

Hint:
- Write a function to calculate euclidean distance from each house (neighborhood) to the school and to the hospital.
- Divide your dataset into houses close and far from either a hospital or school.
- Choose the propper test and, with 5% significance, comment your findings.
 

In [17]:
def euclidean_distance(lon1, lat1, lon2, lat2):
    return np.sqrt((lon2 - lon1)**2 + (lat2 - lat1)**2)

In [19]:
# School coordinates
school_lon = -118
school_lat = 34

# Hospital coordinates
hospital_lon = -122
hospital_lat = 37


df["distance_school"] = euclidean_distance(
    df["longitude"],
    df["latitude"],
    school_lon,
    school_lat
)

df["distance_hospital"] = euclidean_distance(
    df["longitude"],
    df["latitude"],
    hospital_lon,
    hospital_lat
)

distance_school < 0.50 OR distance_hospital < 0.50

In [20]:
df["close"] = (
    (df["distance_school"] < 0.50) |
    (df["distance_hospital"] < 0.50)
)

In [21]:
close_houses = df[df["close"] == True]["median_house_value"]

far_houses = df[df["close"] == False]["median_house_value"]

In [22]:
print("Average price - Close:", close_houses.mean())
print("Average price - Far:", far_houses.mean())

Average price - Close: 246951.98213501245
Average price - Far: 180678.44105790975


In [23]:
t_stat, p_two_tailed = ttest_ind(
    close_houses,
    far_houses,
    equal_var=False
)

# Convert to one-tailed p-value
p_one_tailed = p_two_tailed / 2

print("t-statistic:", t_stat)
print("one-tailed p-value:", p_one_tailed)

t-statistic: 37.992330214201516
one-tailed p-value: 1.5032478884295899e-301


An independent two-sample t-test (Welch's t-test) was performed to compare the median house values of neighborhoods close to a school or hospital with those farther away. The alternative hypothesis stated that houses close to these facilities have higher prices. The test resulted in a t-statistic of 37.99 and a one-tailed p-value of 1.50e-301, which is below the 5% significance level. Therefore, we reject the null hypothesis and conclude that houses close to a school or hospital are significantly more expensive than houses farther away.